test/test_constraints.py

```diff
@@ -69,3 +70,106 @@ def test_unknown_labels_not_in_chunks(num_chunks, chunk_size):
                               random_state=SEED)
 
   assert np.all(chunks[labels < 0] < 0)
+
+
+def test_generate_knntriplets():
+  """Toy example validation of knn triplets construction"""
+  k = 1
+  X = np.array([[0, 0], [2, 2], [4, 4], [8, 8], [16, 16], [32, 32], [64, 64],
+                [128, 128], [256, 256]])
+  y = np.array([1, 1, 1, 2, 2, 2, 3, 3, 3])
+
+  T_test = np.array([[0, 1, 3], [1, 0, 3], [2, 1, 3], [3, 4, 2], [4, 3, 2],
+                     [5, 4, 2], [6, 7, 5], [7, 6, 5], [8, 7, 5]])
+  T = Constraints(y).generate_knntriplets(X, k, k)
+
+  assert np.array_equal(sorted(T.tolist()), sorted(T_test.tolist()))
+
+
+@pytest.mark.parametrize("k_genuine, k_impostor, T_test",
+                         [(2, 3,
+                          [[0, 1, 3], [0, 1, 4], [0, 1, 5], [0, 2, 3],
+                           [0, 2, 4], [0, 2, 5], [1, 0, 3], [1, 0, 4],
+                           [1, 0, 5], [1, 2, 3], [1, 2, 4], [1, 2, 5],
+                           [2, 0, 3], [2, 0, 4], [2, 0, 5], [2, 1, 3],
+                           [2, 1, 4], [2, 1, 5], [3, 4, 0], [3, 4, 1],
+                           [3, 4, 2], [3, 5, 0], [3, 5, 1], [3, 5, 2],
+                           [4, 3, 0], [4, 3, 1], [4, 3, 2], [4, 5, 0],
+                           [4, 5, 1], [4, 5, 2], [5, 3, 0], [5, 3, 1],
+                           [5, 3, 2], [5, 4, 0], [5, 4, 1], [5, 4, 2]]),
+                          (2, 2,
+                          [[0, 1, 3], [0, 1, 4], [0, 2, 3], [0, 2, 4],
+                           [1, 0, 3], [1, 0, 4], [1, 2, 3], [1, 2, 4],
+                           [2, 0, 3], [2, 0, 4], [2, 1, 3], [2, 1, 4],
+                           [3, 4, 1], [3, 4, 2], [3, 5, 1], [3, 5, 2],
+                           [4, 3, 1], [4, 3, 2], [4, 5, 1], [4, 5, 2],
+                           [5, 3, 1], [5, 3, 2], [5, 4, 1], [5, 4, 2]]),
+                          (1, 3,
+                          [[0, 1, 3], [0, 1, 4], [0, 1, 5], [1, 0, 3],
+                           [1, 0, 4], [1, 0, 5], [2, 1, 3], [2, 1, 4],
+                           [2, 1, 5], [3, 4, 0], [3, 4, 1], [3, 4, 2],
+                           [4, 3, 0], [4, 3, 1], [4, 3, 2], [5, 4, 0],
+                           [5, 4, 1], [5, 4, 2]]),
+                          (1, 2,
+                          [[0, 1, 3], [0, 1, 4], [1, 0, 3], [1, 0, 4],
+                           [2, 1, 3], [2, 1, 4], [3, 4, 1], [3, 4, 2],
+                           [4, 3, 1], [4, 3, 2], [5, 4, 1], [5, 4, 2]])])
```

Comment:
Could you also small test edges cases where there is not enough points? typically (1, 4) and (3, 1)?
the outputs should be the same as for (1, 3) and (2, 1) respectively

(you could remove (2,3), which is lengthy)

Reply:
Nevermind about (2,3), it is useful at it corresponds to the case where both values of k are maximum

Reply:
No problem! I will put (2,3) with over the edge cases (3, 3), (2, 4), (3, 4) which should all have the same result.